nomad/state/state_store.go

```diff
@@ -139,6 +146,10 @@ func (a *streamACLDelegate) TokenProvider() stream.ACLTokenProvider {
 
 // NewStateStore is used to create a new state store
 func NewStateStore(config *StateStoreConfig) (*StateStore, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
```

Comment:
We don't have very good existing patterns around parsing configuration in a centralized place but doing validation all the way down here in `NewStateStore` seems like it's "late". We've already started RPC handlers, Vault and Consul clients, etc, etc. by the time we instantiate the state store. Wouldn't it make more sense to validate the config when we parse it?

Reply:
I agree in general yep, and that is already being done [here in agent.go](https://github.com/devashishTaneja/nomad/blob/f32dffc136107e493cb4ea5d0c3cc3fd26779306/command/agent/agent.go#L346-L351) from the previous PR, but the bug arose from a subsequent config hand-off that was missed along the way.

This error _should_ not get hit, due to the config hand-off being filled in now, so this is just an extra guardrail around a sensitive area that uses the value to protect against potential future regression, i.e. make noise in tests if someone accidentally breaks it somehow.

Reply:
Fair enough. It's cheap and not in a hot path.